### DISCUSSION TITLE
How to use two sevenseg HT16K33s to display two sensors (with C->F conversion)?

### DISCUSSION THREAD

**Discussion by markfickett** [arendst/Tasmota#20202] [Q&A] [answered]
I have two [Adafruit HT16K33 I2C backpacks](https://www.adafruit.com/product/879) wired up to a [HUZZAH ESP8266 dev board](https://templates.blakadder.com/adafruit_HUZZAH.html), successfully reading from two [MAX6675 Thermocouple breakouts](https://www.adafruit.com/product/269).

I'd like to show one thermocouple's value on each 7-segment display (as well as sending the values to HA for logging). However, I'm having trouble:

* showing a sensor value on the display
* showing anything on the second display

One display has the default I2C address of 0x70, the other has [A0 shorted so it uses 0x71](https://learn.adafruit.com/adafruit-led-backpack/changing-i2c-address#changing-addresses-706265). It's the second one with A0 shorted that is actually showing something.

Some things that are working:

* Compiling with `USE_DISPLAY` and `USE_DISPLAY_SEVENSEG` and (for the thermocouples) `USE_MAX31855`
* Custom `DisplayText` commands like `DisplayText f98.6` listed [in xdsp_11_sevenseg.ino](https://github.com/tthk/Tasmota/blob/development/tasmota/xdsp_11_sevenseg.ino#L132) and updated [in this PR](https://github.com/arendst/Tasmota/pull/8900) 
* Sensor data is showing like this in the console: `tele/tasmota_30EF0B/SENSOR = {"Time":"2023-12-11T00:54:59","MAX31855-0":{"Temperature":22.8,"ReferenceTemperature":23.6,"Error":0},"MAX31855-1":{"Temperature":22.5,"ReferenceTemperature":22.9,"Error":0},"TempUnit":"C"}` . And I can see both thermocouple values in HomeAssistant.

Some other resources I've looked at:

* [Common-anode 7-segment to show variables](https://github.com/arendst/Tasmota/discussions/14451) discussion, the `%value%` style commands are not working for me
* [Project report of using two I2C e-paper displays](https://github.com/arendst/Tasmota/discussions/13161) but I'm not sure how it would map to what I'm doing if at all
* [7-segment specific commands](https://tasmota.github.io/docs/TM1650/#commands-and-usage) which I gather are different for TM1650 and other related 7-segment displays
* [Resource files for displays](https://tasmota.github.io/docs/Displays/#descriptor-file) I haven't really dug into these, hoping I can get by with more default settings
* [Docs ticket requesting documentation for these I2C displays](https://github.com/tasmota/docs/issues/257)

Here's how it looks:

![overview](https://github.com/arendst/Tasmota/assets/307950/2a3ac6dc-a557-4525-a7e4-b1cbabe2f908)

![backs](https://github.com/arendst/Tasmota/assets/307950/edc73cc1-9722-461a-9bb1-478eae7d1236)

![configuration](https://github.com/arendst/Tasmota/assets/307950/7564bbdb-faa0-432a-a7fa-076a9728694d)

Thanks in advance for any help!

**Comment by barbudor** [comment]
I don't think Tasmota can manage 2 displays simultaneously
At least not natively
You probably have to write a separated driver for the other one using either Scripting-Language or Berry (later available only on ESP32 family)

**Reply by sfromis** [reply, marked as answer]
While I have no experience with this, the comments clearly suggest that up to 8 seven-segment displays are supported, with a command code in `DisplayText` to select which display is addressed.
https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h#L742-L750

**Reply by markfickett** [reply]
Thanks, so it looks like the default `0x70` address is not getting seen because `MTX_ADDRESS1` is `0x71`, and if I short my 2nd display to use `0x74` then I should be able to use `DisplayText [y2]text` with it. I'll give that a shot.

**Reply by markfickett** [reply]
Confirmed that after shorting the other display so it uses `0x74`, I can address it with commands like `DisplayText [y1]f91.4` (and the original one with `0x71` I can address with `DisplayText [y0]f72.1`).

Here's how it looks:

![addresses](https://github.com/arendst/Tasmota/assets/307950/c8bcbfe6-3f08-4bbc-9714-ff40f18fcf22)

![numbers](https://github.com/arendst/Tasmota/assets/307950/0a9400b8-ca7e-472e-85b1-3bc7a1d10e54)

**Comment by sfromis** [comment]
Assuming that the Tasmota support for multiple HTK16K33 works as suggested, making suggestions for how to do things differently gets hard without any details of what you're actually trying, like how you tried addressing the 2nd display, and what you mean by opaque statements like "the `%value%` style commands are not working for me".

For 7-segment displays, I do not think that you should even look in the direction of how it works for e-paper displays or using display descriptor files, as the simple segment displays do not provide such capabilities.

When it comes to C/F, Tasmota does provide `SetOption8` to select which units temperatures are to be represented as.

**Reply by markfickett** [reply]
It probably sounds opaque because I'm really not sure what I should be doing (: but thanks for taking a look.

What I'm looking for is a way to get the sensor value and put it into a `DisplayText` command for my seven segment display. [This other discussion](https://github.com/arendst/Tasmota/discussions/14451) had suggested some rules like `ON tele-BME280#Temperature DO DisplayText [....]Temp: %value% C ENDON` which is where I got the `%value%` from, but I haven't figured out how to make that work. If there's documentation for using variables in rules and you could point me to it, or if you would be willing to help with an example that fits my sensors, that would be awesome.

**Reply by sfromis** [reply]
In `ON tele-BME280#Temperature DO DisplayText [....]Temp: %value% C ENDON` 
`%value%` is simply the value from the trigger, in this case the temperature value, as mentioned [here](https://tasmota.github.io/docs/Rules/#rule-command); scroll down for more available variables.

What you can do in triggers is [here](https://tasmota.github.io/docs/Rules/#rule-trigger); pay special attention to the part about JSON-based triggers, where you have to match the object levels in the JSON you already get for the sensors, similar to the example with BME280.

When you create a rule set, you need to make sure to also enable the rule set, both of this is using the [Rule command](https://tasmota.github.io/docs/Commands/#rules).

More generally, it works better if you post exactly what/how you're trying to make it work, as this makes it easier to spot details which were off. And with rules, the thing is that you can expect nothing to work, and no error messages, until you get every detail right.

**Reply by markfickett** [reply]
Thanks, the [Rule docs](https://tasmota.github.io/docs/Rules/) were what I needed! I didn't realize that `%value%` was something specific to rules, now I see how that works.

The ruleset I am using for now is:

```
Rule1
  ON tele-MAX31855-1#Temperature$!null DO DisplayText[y0]f%value% ENDON
  ON tele-MAX31855-1#Error>0 DO Backlog DisplayText[y0x0]r121[x1]r80[x2]r0[x3]r80[x4]r0 ; Delay 1 ; DisplayText[y0]%value% ENDON
  ON tele-MAX31855-0#Temperature$!null DO DisplayText[y1]f%value% ENDON
  ON tele-MAX31855-0#Error>0 DO Backlog DisplayText[y1x0]r121[x1]r80[x2]r0[x3]r80[x4]r0 ; Delay 1 ; DisplayText[y1]%value% ENDON
BREAK
```

And [SetOption8 1](https://tasmota.github.io/docs/Commands/#setoptions) did the trick for C->F, thanks. Initially I had added the sensors in HomeAssistant when they were in C but HA converted them to F; then when I switched Tasmota to F my values in HA got messed up. But deleting the HA entity and letting HA autodiscover it again let it pick up the right units.

**Comment by lalo-uy** [comment]
This are my rules to display 2 termps on a single 8 digit display:

First keep value on var1 rounded up.
ON Tele-DS18B20-1#Temperature DO Backlog Var1 %value% ; Add1 0.5;  ENDON

Then display secon temp with deg symbol and the first rounded value wit the
same symbol
ON Tele-DS18B20-2#Temperature DO Backlog DisplayText %value%^
;DisplayNumberNC %Var1%,4,0,3; Displaytextnc ^,7;  ENDON

**Reply by markfickett** [reply]
Thanks for the example rules!
